backends/dpdk/midend.cpp

```diff
@@ -0,0 +1,205 @@
+#include "midend/removeAssertAssume.h"
+#include "midend/removeLeftSlices.h"
+#include "midend/removeMiss.h"
+#include "midend/removeParameters.h"
+#include "midend/removeSelectBooleans.h"
+#include "midend/removeUnusedParameters.h"
+#include "midend/replaceSelectRange.h"
+#include "midend/simplifyKey.h"
+#include "midend/simplifySelectCases.h"
+#include "midend/simplifySelectList.h"
+#include "midend/tableHit.h"
+#include "midend/validateProperties.h"
+#include "options.h"
+
+namespace DPDK {
+
+/**
+This class implements a policy suitable for the ConvertEnums pass.
+The policy is: convert all enums that are not part of the psa.
+Use 32-bit values for all enums.
+Also convert PSA_PacketPath_t to bit<32>
+*/
+class PsaEnumOn32Bits : public P4::ChooseEnumRepresentation {
+    cstring filename;
+
+    bool convert(const IR::Type_Enum *type) const override {
+        if (type->name == "PSA_PacketPath_t")
+            return true;
+        if (type->srcInfo.isValid()) {
+            auto sourceFile = type->srcInfo.getSourceFile();
+            if (sourceFile.endsWith(filename))
+                // Don't convert any of the standard enums
+                return false;
+        }
+        return true;
+    }
+    unsigned enumSize(unsigned) const override { return 32; }
+ public:
+    explicit PsaEnumOn32Bits(cstring filename) : filename(filename) {}
+};
+
+PsaSwitchMidEnd::PsaSwitchMidEnd(CompilerOptions &options,
+                                 std::ostream *outStream)
+    : MidEnd(options) {
+    auto convertEnums =
+        new P4::ConvertEnums(&refMap, &typeMap, new PsaEnumOn32Bits("psa.p4"));
+    auto evaluator = new P4::EvaluatorPass(&refMap, &typeMap);
+    std::function<bool(const Context *, const IR::Expression *)> policy =
+        [=](const Context *ctx, const IR::Expression *e) -> bool {
+        if (auto mce = findContext<IR::MethodCallExpression>(ctx)) {
+            auto mi = P4::MethodInstance::resolve(mce, &refMap, &typeMap);
+            if (auto em = mi->to<P4::ExternMethod>()) {
+                if (em->originalExternType->getName().name ==
+                    "InternetChecksum") {
+                    if (em->method->getName().name == "add") {
+                        return false;
+                    }
+                }
+            }
+        }
+
+        auto mce = e->to<IR::MethodCallExpression>();
+        if (mce == nullptr)
+            return true;
+        auto mi = P4::MethodInstance::resolve(mce, &refMap, &typeMap);
+        auto em = mi->to<P4::ExternMethod>();
+        if (em == nullptr)
+            return true;
+        if (em->originalExternType->name.name == "Register" ||
+            em->method->name.name == "read")
+            return false;
+        return true;
+    };
+    if (DPDK::PsaSwitchContext::get().options().loadIRFromJson == false) {
+        std::initializer_list<Visitor *> midendPasses = {
+            options.ndebug ? new P4::RemoveAssertAssume(&refMap, &typeMap)
```

Comment:
you may want to double-check with the latest changes in p4test and see if more passes are needed here